File: 1-windowed-aggregation/ingest-api/0-typescript/src/events/events.service.ts
import { randomUUID } from "node:crypto"
import { Inject, Injectable, Logger, OnModuleInit } from "@nestjs/common"
import { ConfigService } from "@nestjs/config"
import { ClientKafka } from "@nestjs/microservices"
import { KafkaConfig } from "../config/kafka.config"
import { PublishClickDto } from "./publish-click.dto"

/**
 * Producer service: builds a time-stamped click envelope and appends it to the
 * Kafka log. The `eventTime` field drives the tumbling-window boundary on the
 * aggregator; `clickId` is the idempotency key for deduplication.
 */
@Injectable()
export class EventsService implements OnModuleInit {
    private readonly logger = new Logger(EventsService.name)

    constructor(
        @Inject("KAFKA_PRODUCER") private readonly kafka: ClientKafka,
        private readonly config: ConfigService,
    ) {}

    /**
     * Connects the producer once when the module boots.
     *
     * Connecting eagerly surfaces broker problems at startup rather than on
     * the first HTTP request.
     */
    async onModuleInit(): Promise<void> {
        await this.kafka.connect()
    }

    /**
     * Appends one click event to the topic and returns an acknowledgement.
     *
     * @param dto Incoming click payload.
     * @returns Acknowledgement with `eventTime` so callers can inspect the window bucket.
     */
    async publish(
        dto: PublishClickDto,
    ): Promise<{ status: string; topic: string; clickId: string; adId: string; eventTime: string }> {
        const kafka = this.config.getOrThrow<KafkaConfig>("kafka")
        // Mint an idempotency key when the caller did not supply one.
        const clickId = dto.clickId ?? randomUUID()
        // eventTime is the wall-clock time of the click, used by the aggregator
        // to assign the event to its tumbling-window bucket.
        const eventTime = new Date().toISOString()
        const envelope = {
            clickId,
            adId: dto.adId,
            payload: dto.payload ?? {},
            eventTime,
        }
        // key = adId pins all clicks of the same ad to one partition, keeping
        // relative ordering; the header carries clickId for fast dedup without
        // parsing the value body.
        this.kafka.emit(kafka.topic, {
            key: dto.adId,
            value: envelope,
            headers: { clickId },
        })
        this.logger.log(
            `Produced click clickId=${clickId} adId=${dto.adId} eventTime=${eventTime}`,
        )
        return { status: "accepted", topic: kafka.topic, clickId, adId: dto.adId, eventTime }
    }
}

File: 0-event-ingestion-kafka/ingest-api/1-java/src/main/java/com/starci/ingest/EventsService.java
package com.starci.ingest;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.header.internals.RecordHeader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

/**
 * Producer service: builds a click envelope and appends it to the Kafka log.
 *
 * <p>The service is deliberately thin — it owns no counting logic. Counting and
 * deduplication live in the consumer; the producer only guarantees the event
 * carries a stable {@code clickId} so the consumer can deduplicate downstream.
 */
@Service
public class EventsService {

    private static final Logger log = LoggerFactory.getLogger(EventsService.class);

    private final KafkaTemplate<String, String> kafka;
    private final ObjectMapper mapper = new ObjectMapper();

    @Value("${kafka.topic}")
    private String topic;

    public EventsService(KafkaTemplate<String, String> kafka) {
        this.kafka = kafka;
    }

    /**
     * Appends one click event to the topic and returns an acknowledgement.
     *
     * @param dto incoming click payload from {@code POST /clicks}
     * @return a map confirming the append only: {@code status, topic, clickId, adId}
     */
    public Map<String, String> publish(PublishClickDto dto) throws Exception {
        // Mint an idempotency key when the caller did not supply one.
        String clickId = (dto.clickId == null || dto.clickId.isBlank())
                ? UUID.randomUUID().toString()
                : dto.clickId;

        Map<String, Object> envelope = new LinkedHashMap<>();
        envelope.put("clickId", clickId);
        envelope.put("adId", dto.adId);
        envelope.put("payload", dto.payload == null ? Map.of() : dto.payload);
        envelope.put("timestamp", Instant.now().toString());
        String value = mapper.writeValueAsString(envelope);

        // key = adId so all clicks of one ad land on the same partition and keep
        // relative order; the clickId header lets the consumer deduplicate
        // without parsing the value body first.
        ProducerRecord<String, String> record = new ProducerRecord<>(topic, dto.adId, value);
        record.headers().add(new RecordHeader("clickId", clickId.getBytes()));
        // Fire-and-forget: we confirm the append, not consumer processing.
        kafka.send(record);

        log.info("Produced click clickId={} adId={}", clickId, dto.adId);
        return Map.of("status", "accepted", "topic", topic, "clickId", clickId, "adId", dto.adId);
    }
}
